File: NeetCode/Bit_Manipulation/Minimum_Array_End/minimum__array__end.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def minEnd(self, n: int, x: int) -> int:
        n -= 1
        xmask = 1
        nmask = 1
        nbits = n.bit_length()
        xbits = x.bit_length()
        totalBits = nbits + xbits
        result = 0

        for _ in range(totalBits):
            currX = x & xmask
            currN = n & nmask

            if currX == 0:
                result |= currN

            else:
                result |= currX
                n <<= 1
            nmask <<= 1
            xmask <<= 1
        return result


if __name__ == "__main__":
    sol = Solution()
    print(sol.minEnd(97, 10000))
    # print(sol.minEnd(3, 2))
    print("Running Solution...")


"""  

100% works and passes, but i want to try this without bin()

class Solution:
    def minEnd(self, n: int, x: int) -> int:
        n -= 1
        xBits = bin(x)[2:]
        nBits = bin(n)[2:]
        xBits = xBits[::-1]
        nBits = nBits[::-1]
        totalBits = len(xBits) + len(nBits)
        xptr, nptr = 0, 0
        result = ""
        for bit in range(totalBits):
            if xptr < len(xBits):
                if xBits[xptr] == "0":
                    if nptr >= len(nBits):
                        result += "0"
                    xptr += 1
                else:
                    result += "1"
                    xptr += 1
                    continue
            if nptr < len(nBits):
                result += nBits[nptr]
                nptr += 1

        return int(result[::-1], 2)

"""

File: NeetCode/Bit_Manipulation/Minimum_Array_End/Minimum_Array_End.cpp
// Minimum_Array_End.cpp
//
// Compile locally with:
// clang++ Minimum_Array_End.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    long long minEnd(int n, int x)
    {
        long long k = static_cast<long long>(n) - 1;
        long long xmask(1), kmask(1), result(0), currX(0), currK(0);
        int xBits(0), tmpX(x), kBits(0), tmpK(k);
        while (tmpX)
        {
            xBits++;
            tmpX = tmpX >> 1;
        }
        while (tmpK)
        {
            kBits++;
            tmpK = tmpK >> 1;
        }
        int totalbits = xBits + kBits;
        for (int bit = 0; bit < totalbits; bit++)
        {
            currX = x & xmask;
            currK = k & kmask;

            if (currX == 0)
            {
                result = result | currK;
            }
            else
            {
                result = result | currX;
                k = k << 1;
            }
            kmask = kmask << 1;
            xmask = xmask << 1;
        }
        return result;
    }
};

int main()
{
    cout << "Running Minimum_Array_End.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
